README.MD

```diff
@@ -0,0 +1,372 @@
+|------------|------------|-------------|-------------------------------------|
+| CreateJob  | ALLOW      | DENY        | Start new jobs with resource limits |
+| GetJob     | ALLOW      | ALLOW       | Retrieve job details and status     |
+| StopJob    | ALLOW      | DENY        | Terminate running jobs              |
+| StreamLogs | ALLOW      | ALLOW       | Real-time log streaming             |
+
+## Security Architecture
+
+### Certificate-Based Role Management
+
+#### Role Types and Detection
+
+- **Admin Role**: Full system access with job management capabilities
+    - **Detection Methods**:
+        - `OU=Admin` in certificate subject
+        - `admin.*.local` in Subject Alternative Names
+        - "admin" substring in Common Name
+    - **Permissions**: All operations (CRUD + streaming)
+
+- **Viewer Role**: Read-only access for monitoring
+    - **Detection Methods**:
+        - `OU=Viewer` in certificate subject
+        - `viewer.*.local` in Subject Alternative Names
+        - "viewer" substring in Common Name
+    - **Permissions**: Get, List, Stream operations only
+
+#### Certificate Infrastructure
+
+#### TLS Configuration
+
+- **Version**: TLS 1.3 minimum
+- **Authentication**: Mutual TLS with mandatory client certificates
+- **Key Strengths**: `4096-bit` CA, `2048-bit` clients keys (admin/viewer)
+- **Certificate Lifetime**: `365` days with renewal processes
+
+### Authorization Flow
+
+![Authorization Flow](docs/job-worker-authorization-flow.drawio.png)
+
+## Data Flow Architecture
+
+### Job Lifecycle
+
+![Job Lifecycle](docs/job-worker-job-lifecycle.drawio.png)
+
+### Real-Time Log Streaming
+
+![Log Streaming Flow](docs/job-worker-log-streaming-flow.drawio.png)
+
+**Pub-Sub Implementation Details**:
+
+The system implements an in-memory pub-sub mechanism without external message brokers:
+
+1. **Task-Level Subscribers**: Each Task (Task: is a wrapper around the job) maintains its own subscriber list
+
+2. **Subscription Management**:
+    - **Subscribe**: Creates buffered channel, adds to subscriber list
+    - **Unsubscribe**: Removes channel from list, closes channel
+    - **Auto-cleanup**: Detects stalled subscribers (`5s` timeout), removes automatically
+
+3. **Message Flow**:
+    - Process Output → `io.Writer` publish (Update) → Subscriber Channels
+    - **Update:** Contains the:
+      - JobId
+      - LogChunk `[]byte`
+      - Status
+
+4. **Delivery Guarantees**:
+    - **Best Effort**: Non-blocking sends with timeout
+    - **Buffered**: Each subscriber has `50` buffer
+    - **Cleanup**: Slow/dead subscribers automatically removed
+    - **Multiple Subscribers**: Same log chunk delivered to all active subscribers
+
+5. **Thread Safety**:
+    - Copy-on-read pattern for subscriber list during publishing
+    - Atomic operations for subscriber management
+
+6. **Performance Characteristics**:
+    - **Memory**: Bounded by subscriber count × buffer size
```

Comment:
can we improve this?

Reply:
sorry for the late reply. that's not what i meant. cgroup is for the job not for the client.

for clients, no matter how many subscribers to stream the job, they are all reading the same content at the end of the day. so i am just asking if any optimization can be done to avoid `50` buffer per subscriber.

it's just a question though, you don't have to derail from your current design.

Reply:
I don't quite understand the reasoning behind correlating the size of a job's output buffer to the amount of memory the job is allowed to use when running. I don't think we can assume there's any correlation between those two things. You can have a job that is very intensive but produces little output, or a job that uses little memory but is very chatty.